crates/iceberg-catalog/src/implementations/postgres/task_queues.rs

```diff
@@ -0,0 +1,385 @@
+                        idempotency_key)
+        VALUES ($1, $2, 'pending', $3, $4)
+        ON CONFLICT ON CONSTRAINT unique_idempotency_key DO NOTHING
+        RETURNING task_id"#,
+        task_id,
+        task_name,
+        parenet_task_id,
+        idempotency_key
+    )
+    .fetch_optional(conn)
+    .await
+    .map_err(|e| e.into_error_model("fail".into()))?;
+
+    if task_handle.is_none() {
+        tracing::info!(
+            "Task already exists with idempotency key: {}",
+            idempotency_key
+        );
+        return Ok(task_id);
+    }
+    Ok(task_id)
+}
+
+pub(crate) async fn queue_delete(
+    conn: &mut PgConnection,
+    DeleteInput {
+        entity_id,
+        location,
+        warehouse_id,
+        parent_task,
+    }: DeleteInput,
+    task_name: &'static str,
+) -> Result<Uuid, IcebergErrorResponse> {
+    tracing::info!(
+        "Queuing delete for '{location}' under entity_id: {}",
+        entity_id
+    );
+
+    let idempotency_key = Uuid::new_v5(&warehouse_id, location.as_bytes());
+
+    let task_id = queue_task(conn, task_name, parent_task, idempotency_key).await?;
+
+    let it = sqlx::query!(
+        r#"
+        INSERT INTO deletions (task_id, entity_id, location, warehouse_id)
+        VALUES ($1, $2, $3, $4)
+        ON CONFLICT ON CONSTRAINT unique_location_per_warehouse DO NOTHING
+        RETURNING task_id
+        "#,
+        task_id,
+        entity_id,
+        location,
+        *warehouse_id,
+    )
+    .fetch_optional(conn)
+    .await
+    .map_err(|e| e.into_error_model("fail".into()))?;
+
+    match it {
+        Some(row) => tracing::info!("Queued delete task: {:?}", row),
+        None => {
+            tracing::info!("Delete task already exists for location: {}", location);
+        }
+    }
+    Ok(task_id)
+}
+
+async fn record_failure(
+    conn: &PgPool,
+    id: Uuid,
+    n_retries: i32,
+    details: String,
+) -> Result<(), IcebergErrorResponse> {
+    let _ = sqlx::query!(
+        r#"
+        WITH cte as (
+            SELECT attempt >= $2 as should_fail
+            FROM task
+            WHERE task_id = $1
+        )
+        UPDATE task
+        SET status = CASE WHEN (select should_fail from cte) THEN 'failed'::task_status ELSE 'pending'::task_status END,
+            last_error_details = $3
+        WHERE task_id = $1
+        "#,
+        id,
+        n_retries,
+        details
+    )
+    .execute(conn)
+    .await.map_err(|e| e.into_error_model("fail".into()))?;
+    Ok(())
+}
+
+#[tracing::instrument]
+async fn pick_task(
+    conn: &mut PgConnection,
+    name: &'static str,
+) -> Result<Option<Task>, IcebergErrorResponse> {
+    let x = sqlx::query_as!(
+        Task,
+        r#"
+    WITH updated_task AS (
+        SELECT task_id
+        FROM task
+        WHERE status = 'pending' AND task_name = $1
+        FOR UPDATE SKIP LOCKED
+        LIMIT 1
+    )
+    UPDATE task
+    SET status = 'running', picked_up_at = $2, attempt = task.attempt + 1
```

Comment:
Maybe I missed it - but I think we need a max-run-duration mechanism that resets the state from running to pending if a task is still running after x minutes.
Otherwise, on hard-kills of workers, we might end up having running tasks that are never consolidated.

Reply:
there's a to-do in the fn that fetches a task for this